modules/libtools/netstat.cc

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2019 Cloudius Systems, Ltd.
+ *
+ * This work is open source software, licensed under the terms of the
+ * BSD license as described in the LICENSE file in the top-level directory.
+ */
+
+#include <bsd/porting/netport.h>
+#include <bsd/porting/route.h>
+
+#include <bsd/sys/sys/sysctl.h>
+#include <bsd/sys/sys/socket.h>
+#include <bsd/sys/sys/socketvar.h>
+#include <bsd/sys/netinet/in.h>
+#include <bsd/sys/netinet/ip_var.h>
+#include <bsd/sys/netinet/in_pcb.h>
+#include <bsd/sys/netinet/tcp_var.h>
+#include <bsd/sys/netinet/tcp_fsm.h>
+#include <bsd/sys/netinet/udp.h>
+#include <bsd/sys/netinet/udp_var.h>
+
+#include <bsd/sys/compat/linux/linux.h>
+#include <bsd/sys/compat/linux/linux_socket.h>
+
+#include <errno.h>
+#include <sys/sysctl.h>
+
+#include "netstat.hh"
+
+
+namespace osv {
+
+static inline void format_ip(std::string &str, int family, void *addr)
+{
+    char ipstr[INET6_ADDRSTRLEN];
+    if (inet_ntop(family, addr, ipstr, sizeof(ipstr)) == 0)
+        str = "????";
+    else
+        str = ipstr;
+}
+
+static inline void format_tcp_state(std::string &str, int state)
+{
+    struct sn {
+        int state;
+        const char *name;
+    } states [] = { 
+        { TCPS_CLOSED, "CLOSED" },
+        { TCPS_LISTEN, "LISTEN" },
+        { TCPS_SYN_SENT, "SYN_SENT" },
+        { TCPS_SYN_RECEIVED, "SYN_RECEIVED" },
+        { TCPS_ESTABLISHED, "ESTABLISHED" },
+        { TCPS_CLOSE_WAIT, "CLOSE_WAIT" },
+        { TCPS_FIN_WAIT_1, "FIN_WAIT_1" },
+        { TCPS_CLOSING, "CLOSING" },
+        { TCPS_LAST_ACK, "LAST_ACK" },
+        { TCPS_FIN_WAIT_2, "FIN_WAIT_2" },
+        { TCPS_TIME_WAIT, "TIME_WAIT" }
+    };
+
+    for (auto&& s : states) {
+        if (s.state == state) {
+            str = s.name;
+            return;
+        } 
+    }
+
+    char buf[16];
+    snprintf(buf, sizeof(buf), "state=%d?", state);
+    str = buf;
+}
+
+static void format_proto(std::string &str, int proto, bool v6)
+{
+    if (proto == IPPROTO_TCP)
+        str = v6 ? "tcp6" : "tcp";
+    else if (proto == IPPROTO_UDP)
+        str = v6 ? "udp6" : "udp";
+    else {
+        char buf[16];
+        snprintf(buf, sizeof(buf), "proto=%d?", proto);
+        str = buf;
+    }
+}
+
+static int foreach_protocol_pcb(int proto, int proto_ctl, pcb_stats_fun fun)
+{
+    struct xinpgen *inpg;
+    char *sp, *ep, *cp, *np;
+    size_t needed = -1;
+    int mib[4];
+    char ipstr[INET6_ADDRSTRLEN];
+
+    mib[0] = CTL_NET;
+    mib[1] = PF_INET;
+    mib[2] = proto;
+    mib[3] = proto_ctl;
+
+    if (osv_sysctl(mib, 4, NULL, &needed, NULL, 0) < 0) {
+        fprintf(stderr, "PCBLIST sysctl failed to get buffer size\n");
+        return errno;
+    }
+
+    sp = new char[needed];
```

Comment:
Surely you mean `std::vector<char>(needed)`?
At least that way, you don't have to worry about deleting `sp` on exit.